scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1089,24 +1089,74 @@ final case class MergedTypedPipe[T](left: TypedPipe[T], right: TypedPipe[T]) ext
         case (pipe, cnt) => pipe.flatMap(List.fill(cnt)(_).iterator)
       }
       .map(_.toPipe[U](fieldNames)(flowDef, mode, setter)) // linter:ignore
-      .toList
+      .toVector
 
     if (merged.size == 1) {
       // there is no actual merging here, no need to rename:
       merged.head
     } else {
-      merged.reduce[Pipe] {
-        case (left, right) =>
-          // special handling for cases where one side of the hashjoin is merged
-          // with the hashjoin result. Cascading no longer allows it,
-          // so we add a checkpoint to the join result as a workaround
-          if (RichPipe.isHashJoinedWithPipe(left, right))
-            new Merge(RichPipe.assignName(new Checkpoint(left)), RichPipe.assignName(right))
-          else if (RichPipe.isHashJoinedWithPipe(right, left))
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(new Checkpoint(right)))
-          else
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(right))
+      mergeAvoidingHashes(merged)
+    }
+  }
+
+  /**
+   * Special handling for cases where one side of the hashjoin is merged
+   *  with the hashjoin result. Cascading no longer allows it (as of 3.0),
+   *  so we insert checkpoints and/or intermediate merge stages as appropriate
+   *
+   * @param pipes a list of pipes possibly colliding with each other
+   * @return an updated list of pipes, which can safely be merged
+   */
+  protected def mergeAvoidingHashes(pipes: Vector[Pipe]): Pipe = {
+    require(!pipes.isEmpty)
+
+    // we make use of the fact that the pipe merge operation is not just associative but also commutative
+    val (colliding, uncolliding) = pipes.partition(p => pipes.exists(o => (o != p) && RichPipe.isHashJoinedWithPipe(p, o)))
+
+    val (innerColliding, innerUncolliding) = colliding.partition(p =>
+      pipes.exists(o => (o != p) && (RichPipe.isHashJoinedWithPipe(p, o) || RichPipe.isHashJoinedWithPipe(o, p))))
+    /* inneruncolliding pipes collide with some pipes in the uncolliding set, but don't collide with one another.
+     It is fine to lump them and merge them together before merging with the 'uncolliding' set.
+
+      innercolliding pipes collide with one another and must each be checkpointed before use in the general merge.
+     */
+    val safedInnerColliding = innerColliding.map(new Checkpoint(_))
+    val safedInnerUncolliding = if (innerUncolliding.isEmpty) Nil else Seq(checkpointOrMergeAvoidingNameClashes(innerUncolliding))
+
+    mergeAvoidingNameClashes(safedInnerColliding ++ safedInnerUncolliding ++ uncolliding)
+  }
+
+  /**
+   * Ensure there is either a merge or a checkpoint resulting from the incoming pipes.
+   *
+   * @param pipes
+   * @return
+   */
+  protected def checkpointOrMergeAvoidingNameClashes(pipes: Seq[Pipe]): Pipe = {
+    if (pipes.size == 1) {
+      new Checkpoint(pipes.head)
```

Comment:
Turns out the Checkpoint is always necessary, which dissolves the need for ```checkpointOrMergeAvoidingNameClashes()``` entirely (I had cases where running without the checkpoint around the inner merge worked, but I couldn't always reliably rule out a violation in the planner's RemoveMalformedHashJoins rule)